### PR TITLE
Constituent Mods

### PR DESCRIPTION
**Pull request by peverwhee**
Summary of changes
* Split up register constituents into register and initialize (need to do separately because we need the grid setup to do the initialization but need the num_constituents to do the grid setup
* Add constituent property initialized_in_physics so we can query for that and conditionally not read the var from file
* Add module-level variable to ccpp_hash_table to avoid dangling pointer issues (when the pointer passed into hash_iterator_initialize goes out of scope, we lose that memory)

### CODE REVIEW

**Comment by gold2718** [comment]
I am trying to digest these changes.
One thing that would help would be to "fix" test/advection_test so that it works with the framework changes.

**Comment by peverwhee** [comment]
@gold2718 will do!

**Comment by gold2718** [comment]
./test/advection_test/run_test should run to completion without errors.
To check your setup, ./test/capgen_test/run_test should also run (not to mention all the unit tests).

**Comment by peverwhee** [comment]
Apologies for the delay - all tests pass now

File: test/advection_test/cld_ice.F90 [hunk 1]
```diff
@@ -19,15 +19,15 @@ MODULE cld_ice
    !> \section arg_table_cld_ice_run  Argument Table
    !! \htmlinclude arg_table_cld_ice_run.html
    !!
-   subroutine cld_ice_run(ncol, timestep, temp, qv, ps, cld_ice,              &
+   subroutine cld_ice_run(ncol, timestep, temp, qv, ps, cld_ice_array,              &
```
Comment (gold2718) [inline test/advection_test/cld_ice.F90:22]: What is the point / necessity of / for this change?

File: test/advection_test/cld_liq.F90 [hunk 2]
```diff
@@ -16,7 +16,7 @@ MODULE cld_liq
    !> \section arg_table_cld_liq_run  Argument Table
    !! \htmlinclude arg_table_cld_liq_run.html
    !!
-   subroutine cld_liq_run(ncol, timestep, tcld, temp, qv, ps, cld_liq,        &
+   subroutine cld_liq_run(ncol, timestep, tcld, temp, qv, ps, cld_liq_array,        &
```
Comment (gold2718) [inline test/advection_test/cld_liq.F90:19]: Same question, why was this necessary?

File: test/advection_test/test_host.F90 [hunk 3]
```diff
@@ -223,8 +222,9 @@ subroutine test_host(retval, test_suites)
        use test_host_mod,      only: init_data, compare_data
        use test_host_mod,      only: ncols, pver
        use test_host_ccpp_cap, only: test_host_ccpp_register_constituents
+       use test_host_ccpp_cap, only: test_host_ccpp_initialize_constituents
        use test_host_ccpp_cap, only: test_host_ccpp_number_constituents
-       use test_host_ccpp_cap, only: test_host_advected_constituents
+       use test_host_ccpp_cap, only: test_host_constituents_array
```
Comment (gold2718) [inline test/advection_test/test_host.F90:227]: Is this really the same test or did you change it to test something else?
Can you explain why the changes are needed and how they still test the same functionality?

File: src/ccpp_hash_table.F90 [hunk 4]
```diff
@@ -59,8 +59,11 @@ module ccpp_hash_table
       procedure :: next       => hash_iterator_next_entry
       procedure :: valid      => hash_iterator_is_valid
       procedure :: value      => hash_iterator_value
+      procedure :: finalize   => hash_iterator_finalize
    end type ccpp_hash_iterator_t
 
+   !! Module-level hash table variable
+   type(ccpp_hash_table_t), private, allocatable, target :: hash_table_module
```
Comment (gold2718) [inline src/ccpp_hash_table.F90]: What is the point of (i.e., necessity for) this variable? How will this work when there are multiple hash tables flying around (e.g., history plus (potentially multiple) CCPP suite(s)? What about thread safety?

File: src/ccpp_hash_table.F90 [hunk 5]
```diff
@@ -517,4 +522,15 @@ function hash_iterator_value(this) result(val)
...
+   subroutine hash_iterator_finalize(this)
+      ! Deallocate hash_table object
+
+      ! Dummy arguments
+      class(ccpp_hash_iterator_t)    :: this
+
+      nullify(this%hash_table)
+      deallocate(hash_table_module)
+
+   end subroutine hash_iterator_finalize
```
Comment (gold2718) [inline src/ccpp_hash_table.F90]: Is this only necessary because there is a module-level hash table?

File: test/var_action_test/effr_calc.F90 [hunk 6]
```diff
@@ -28,9 +28,9 @@ subroutine effr_calc_run(ncol, nlev, effrr_in, effrl_inout, &
       integer,            intent(out)   :: errflg
       !----------------------------------------------------------------
 
-      real(kind_phys), parameter :: re_qc_min = 2.5 ! microns
-      real(kind_phys), parameter :: re_qc_max = 50. ! microns
-      real(kind_phys), parameter :: re_qi_avg = 75. ! microns
+      real(kind_phys), parameter :: re_qc_min = 2.5E-6 ! 2.5 microns, in meters
+      real(kind_phys), parameter :: re_qc_max = 5.0E-5 ! 50 microns, in meters
+      real(kind_phys), parameter :: re_qi_avg = 7.5E-5 ! 75 microns, in meters
```
Comment (gold2718) [inline test/var_action_test/effr_calc.F90]: I think the changes in this file need to be reverted. This test is supposed to fail until the automatic unit conversion code is in place and working.

File: test/var_action_test/test_host.F90 [hunk 7]
```diff
@@ -329,7 +329,7 @@ subroutine test_host(retval, test_suites)
 
        if (errflg == 0) then
           ! Run finished without error, check answers
-          if (compare_data()) then
+          if (compare_data(col_start, col_end)) then
```
Comment (gold2718) [inline test/var_action_test/test_host.F90]: I think the changes in this file need to be reverted. This test is supposed to fail until the automatic unit conversion code is in place and working.

File: test/var_action_test/test_host_mod.F90 [hunk 8]
```diff
@@ -30,7 +30,9 @@ subroutine init_data()
 
   end subroutine init_data
 
-  logical function compare_data()
+  logical function compare_data(col_start, col_end)
+    integer,        intent(in) :: col_start
+    integer,        intent(in) :: col_end
```
Comment (gold2718) [inline test/var_action_test/test_host_mod.F90]: I think the changes in this file need to be reverted. This test is supposed to fail until the automatic unit conversion code is in place and working.

File: scripts/metavar.py [hunk 9]
```diff
@@ -332,6 +332,13 @@ def __init__(self, prop_dict, source, run_env, context=None,
                 break
             # end if
         # end for
+        self.__initialized_in_physics = False
+        # Check for variables that are initialized in physics scheme(s)
+        if source.name and source.name[-4:] == 'init' and source.name[-14:] != 'timestep_init' and 'intent' in prop_dict:
+            if prop_dict['intent'] == 'out':
+                self.__initialized_in_physics = True
+            # end if
```
Comment (gold2718) [inline scripts/metavar.py]: Still not convinced this is necessary.

File: scripts/constituents.py [hunk 10]
```diff
@@ -112,7 +112,7 @@ def find_variable(self, standard_name=None, source_var=None,
                 # end for
                 newdims.append(':'.join(new_dnames))
             # end for
-            var = source_var.clone({'dimensions' : newdims}, remove_intent=True,
+            var = source_var.clone({'dimensions' : newdims}, remove_intent=False,
```
Comment (gold2718) [inline scripts/constituents.py:115]: I cannot wrap my head around this change. The constituents dictionary was never intended to be a source of variables for dummy arguments. I need a clear explanation why this is necessary.

File: src/ccpp_constituent_prop_mod.F90 [hunk 11]
```diff
@@ -595,6 +610,23 @@ end subroutine ccp_is_advected
...
+      character(len=*), optional,           intent(out) :: errmsg
+
+      if (this%is_instantiated(errcode, errmsg)) then
+         val_out = this%initialized_in_physics
+      else
+         val_out = .false.
+      end if
+   end subroutine ccp_is_initialized_in_physics
+
+   !#######################################################################
```
Comment (gold2718) [inline src/ccpp_constituent_prop_mod.F90]: Still not convinced this is necessary.

**Review by gold2718 — CHANGES_REQUESTED** [review]
I have a lot of deep (in the code and in the way the framework works) questions and concerns.

File: test/advection_test/cld_ice.F90 (hunk 1 above)
Comment (peverwhee) [reply test/advection_test/cld_ice.F90:22]: I was getting an error complaining about a conflict with this use statement:
> error #6412: A dummy argument name is invalid in this context.   [CLD_ICE]
>       use cld_ice, only: cld_ice_init

File: test/advection_test/test_host.F90 (hunk 3 above)
Comment (peverwhee) [reply test/advection_test/test_host.F90:227]: Changed it to the test_host_constituents_array because that more accurately reflects how we're using the constituent array (though the two arrays are identical)

File: src/ccpp_hash_table.F90 (hunk 4 above)
Comment (peverwhee) [reply src/ccpp_hash_table.F90]: NAG was giving us a dangling pointer error without this, but your point is well taken! This change has been reverted and we're working on testing with an updated version of NAG that hopefully can handle this suuuper complex fortran

File: test/var_action_test/effr_calc.F90 (hunk 6 above)
Comment (peverwhee) [reply test/var_action_test/effr_calc.F90]: reverted

File: src/ccpp_constituent_prop_mod.F90 (hunk 11 above)
Comment (peverwhee) [reply src/ccpp_constituent_prop_mod.F90]: removed

File: scripts/metavar.py (hunk 9 above)
Comment (peverwhee) [reply scripts/metavar.py]: removed

File: scripts/constituents.py (hunk 10 above)
Comment (peverwhee) [reply scripts/constituents.py:115]: This I did keep. I needed to know if a constituent was intent out because the framework needs to throw an error if you try to declare an intent "out" constituent at runtime (or, just any time other than "init" time)

**Comment by peverwhee** [comment]
@gold2718 I've removed most of what I think you were bumping on. I did open a PR to feature/capgen so feel free to express any further concerns there.

**Comment by gold2718** [comment]
Closing this to reverse directions.
